curvlinops/ekfac.py

```diff
@@ -0,0 +1,638 @@
+            backward hook on the module itself which performs the computation.
+            However, this approach breaks down if the output of a layer feeds into an
+            activation with `inplace=True` (see
+            https://github.com/pytorch/pytorch/issues/61519). Hence we use the
+            workaround
+            https://github.com/pytorch/pytorch/issues/61519#issuecomment-883524237, and
+            install a module hook which installs a tensor hook on the module's output
+            tensor, which performs the accumulation of the gradient covariance.
+
+        Args:
+            module: Layer onto whose output a tensor hook to accumulate the corrected
+                eigenvalues will be installed.
+            inputs: The layer's input tensors.
+            output: The layer's output tensor.
+            module_name: The name of the layer in the neural network.
+        """
+        tensor_hook = partial(
+            self._accumulate_corrected_eigenvalues,
+            module=module,
+            module_name=module_name,
+        )
+        output.register_hook(tensor_hook)
+
+    def _accumulate_corrected_eigenvalues(
+        self, grad_output: Tensor, module: Module, module_name: str
+    ):
+        r"""Accumulate the corrected eigenvalues.
+
+        The corrected eigenvalues are computed as
+        :math:`\lambda_{\text{corrected}} = (Q_g^T G Q_a)^2`, where
+        :math:`Q_a` and :math:`Q_g` are the eigenvectors of the input and gradient
+        covariances, respectively, and ``G`` is the gradient matrix. The corrected
+        eigenvalues are used to correct the eigenvalues of the KFAC approximation
+        (EKFAC).
+
+        Updates ``self._corrected_eigenvalues``.
+
+        Args:
+            grad_output: The gradient w.r.t. the output.
+            module: The layer for which corrected eigenvalues will be accumulated.
+            module_name: The name of the layer in the neural network.
+        """
+        g = grad_output.data.detach()
+        batch_size = g.shape[0]
+        if isinstance(module, Conv2d):
+            g = rearrange(g, "batch c o1 o2 -> batch o1 o2 c")
+        g = rearrange(g, "batch ... d_out -> batch (...) d_out")
+
+        # Compute correction for the loss scaling depending on the loss reduction used
+        num_loss_terms = batch_size * self._num_per_example_loss_terms
+        # self._mc_samples will be 1 if fisher_type != FisherType.MC
+        correction = {
+            "sum": 1.0 / self._mc_samples,
+            "mean": num_loss_terms**2
+            / (self._N_data * self._mc_samples * self._num_per_example_loss_terms),
+        }[self._loss_func.reduction]
+
+        # Compute the corrected eigenvalues for the EKFAC approximation
+        param_pos = self._mapping[module_name]
+        aaT_eigenvectors = self._input_covariances_eigenvectors.get(module_name)
+        ggT_eigenvectors = self._gradient_covariances_eigenvectors.get(module_name)
```

Comment:
Better to use `[module_name]` here because this may still pass if the key does not exist?

Reply:
We can only do this for `ggT_eigenvectors` because `aaT_eigenvectors` does not exist when the weight matrix is excluded. I made the change for `ggT_eigenvectors` and added a clarifying comment.